### PR TITLE
RUL-37: Introduce attribute copiers for different attribute types

### PR DESCRIPTION
**Pull request by mmetayer**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | -
| Changelog updated                 | -
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Component/Product/Updater/Copier/HeterogeneousAttributeCopier.php [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+use Akeneo\Tool\Component\StorageUtils\Exception\InvalidPropertyException;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+/**
+ * @copyright 2020 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ *
+ * Copies attribute data from different attribute types
+ */
+class HeterogeneousAttributeCopier implements AttributeCopierInterface
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Updater/Copier/HeterogeneousAttributeCopier.php]: I don't really like the naming, feel free to suggest smth better

File: src/Akeneo/Pim/Enrichment/Component/Product/Updater/Converter/NumberToStringDataConverter.php [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+use Akeneo\Pim\Enrichment\Component\Product\Model\ValueInterface;
+use Akeneo\Pim\Structure\Component\AttributeTypes;
+use Akeneo\Pim\Structure\Component\Model\AttributeInterface;
+use Webmozart\Assert\Assert;
+
+/**
+ * @copyright 2020 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class NumberToStringDataConverter implements ValueDataConverter
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Updater/Converter/NumberToStringDataConverter.php]: final

File: src/Akeneo/Pim/Enrichment/Component/Product/Updater/Copier/HeterogeneousAttributeCopier.php (hunk 1 above)
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Updater/Copier/HeterogeneousAttributeCopier.php]: final

File: src/Akeneo/Pim/Enrichment/Component/Product/Updater/Copier/HeterogeneousAttributeCopier.php [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        $fromLocale = $options['from_locale'];
+        $toLocale = $options['to_locale'];
+        $fromScope = $options['from_scope'];
+        $toScope = $options['to_scope'];
+
+        $this->checkLocaleAndScope($fromAttribute, $fromLocale, $fromScope);
+        $this->checkLocaleAndScope($toAttribute, $toLocale, $toScope);
+
+        $fromValue = $fromEntityWithValues->getValue($fromAttribute->getCode(), $fromLocale, $fromScope);
+        $targetData = null !== $fromValue ? $this->valueDataConverterRegistry->getDataConverter($fromAttribute, $toAttribute)->convert(
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Component/Product/Updater/Copier/HeterogeneousAttributeCopier.php]: `getDataConverter(...)` can  return null. It should not be chained, and null case should be handled

File: tests/back/Pim/Enrichment/Specification/Component/Product/Updater/Converter/StringToStringDataConverterSpec.php [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+use Akeneo\Pim\Enrichment\Component\Product\Updater\Converter\StringToStringDataConverter;
+use Akeneo\Pim\Enrichment\Component\Product\Value\ScalarValue;
+use Akeneo\Pim\Structure\Component\AttributeTypes;
+use Akeneo\Pim\Structure\Component\Model\Attribute;
+use Akeneo\Pim\Structure\Component\Model\AttributeInterface;
+use PhpSpec\ObjectBehavior;
+
+class StringToStringDataConverterSpec extends ObjectBehavior
+{
+    function it_is_a_simple_data_converter()
```
Comment (nmarniesse) [inline tests/back/Pim/Enrichment/Specification/Component/Product/Updater/Converter/StringToStringDataConverterSpec.php]:
```suggestion
    function it_is_initializable()
```

Miss `it_is_a_value_data_converter` function

File: tests/back/Pim/Enrichment/Specification/Component/Product/Updater/Converter/StringToStringDataConverterSpec.php [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        $this->shouldHaveType(StringToStringDataConverter::class);
+    }
+
+    function it_supports_a_combination_of_source_an_target_attributes(
```
Comment (nmarniesse) [inline tests/back/Pim/Enrichment/Specification/Component/Product/Updater/Converter/StringToStringDataConverterSpec.php]:
```suggestion
    function it_supports_a_combination_of_source_and_target_attributes(
```

File: tests/back/Pim/Enrichment/Specification/Component/Product/Updater/Converter/ValueDataConverterRegistrySpec.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        AttributeInterface $sourceAttribute,
+        AttributeInterface $targetAttribute
+    ) {
+        $firstConverter->supportsAttributes($sourceAttribute, $targetAttribute)->willReturn(false);
+        $secondConverter->supportsAttributes($sourceAttribute, $targetAttribute)->willReturn(true);
+
+        $this->getDataConverter($sourceAttribute, $targetAttribute)->shouldReturn($secondConverter);
+    }
+
+    function it_returns_null_if_attributes_are_not_supported(
```
Comment (nmarniesse) [inline tests/back/Pim/Enrichment/Specification/Component/Product/Updater/Converter/ValueDataConverterRegistrySpec.php]: The test name does not match with the test

File: tests/back/Pim/Enrichment/Specification/Component/Product/Updater/Copier/HeterogeneousAttributeCopierSpec.php [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+        $this->supportsAttributes($name, $color)->shouldReturn(true);
+
+        $weight->getType()->willReturn(AttributeTypes::METRIC);
+        $valueDataConverter->supportsAttributes($name, $weight)->willReturn(false);
+
+        $this->supportsAttributes($name, $weight)->shouldReturn(false);
+    }
+
+    function it_copies_data_form_empty_values_without_converting_them(
```
Comment (nmarniesse) [inline tests/back/Pim/Enrichment/Specification/Component/Product/Updater/Copier/HeterogeneousAttributeCopierSpec.php]: for? from?

File: src/Akeneo/Pim/Enrichment/Component/Product/Updater/Copier/HeterogeneousAttributeCopier.php (hunk 3 above)
Comment (nmarniesse) [reply src/Akeneo/Pim/Enrichment/Component/Product/Updater/Copier/HeterogeneousAttributeCopier.php]: Ok I didn't get the `supportsAttributes` method should prevent that. Do as you want ;)

File: src/Akeneo/Pim/Enrichment/Component/Product/Updater/Converter/ValueDataConverter.php [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+{
+    public function supportsAttributes(AttributeInterface $sourceAttribute, AttributeInterface $targetAttribute): bool;
+
+    /**
+     * @param ValueInterface $sourceValue
+     * @param AttributeInterface $targetAttribute
+     *
+     * @return mixed
+     */
+    public function convert(ValueInterface $sourceValue, AttributeInterface $targetAttribute);
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Component/Product/Updater/Converter/ValueDataConverter.php]: IMO it's hard to understand what the method does, specially what is the role of the $targetAttribute and what the method returns.  
But I don't have any better ideas...

File: src/Akeneo/Pim/Enrichment/Component/Product/Updater/Copier/HeterogeneousAttributeCopier.php (hunk 3 above)
Comment (mmetayer) [reply src/Akeneo/Pim/Enrichment/Component/Product/Updater/Copier/HeterogeneousAttributeCopier.php]: Yep, I know it's not ideal, we can try to find a better solution IRL

**Pull request merged**